lib/modules/manager/nuget/config-formatter.ts

```diff
@@ -0,0 +1,118 @@
+import * as hostRules from '../../../util/host-rules';
+import { NugetDatasource } from '../../datasource/nuget';
+import { parseRegistryUrl } from '../../datasource/nuget/common';
+import type { ParsedRegistryUrl } from '../../datasource/nuget/types';
+import type {
+  PackageSourceCredential,
+  PackageSourceMap,
+  Registry,
+} from './types';
+
+export function createNuGetConfigXml(registries: Registry[]): string {
+  let contents = `<?xml version="1.0" encoding="utf-8"?>\n<configuration>\n<packageSources>\n`;
+  let unnamedRegistryCount = 0;
+
+  const credentials: PackageSourceCredential[] = [];
+  const packageSourceMaps: PackageSourceMap[] = [];
+
+  for (const registry of registries) {
+    const registryName =
+      registry.name ?? `Package source ${++unnamedRegistryCount}`;
+    const registryInfo = parseRegistryUrl(registry.url);
+
+    contents += formatPackageSourceElement(registryInfo, registryName);
+
+    const { password, username } = hostRules.find({
+      hostType: NugetDatasource.id,
+      url: registry.url,
+    });
+
+    // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+    if (password || username) {
+      credentials.push({
+        name: registryName,
+        password,
+        username,
+      });
+    }
+
+    if (registry.sourceMappedPackagePatterns) {
+      packageSourceMaps.push({
+        name: registryName,
+        patterns: registry.sourceMappedPackagePatterns,
+      });
+    }
+  }
+
+  contents += '</packageSources>\n';
+
+  if (credentials.length > 0) {
+    contents += '<packageSourceCredentials>\n';
+
+    for (const credential of credentials) {
+      contents += formatPackageSourceCredentialElement(credential);
+    }
+
+    contents += '</packageSourceCredentials>\n';
+  }
+
+  if (packageSourceMaps.length > 0) {
+    contents += '<packageSourceMapping>\n';
+    for (const packageSourceMap of packageSourceMaps) {
+      contents += formatPackageSource(packageSourceMap);
+    }
+    contents += '</packageSourceMapping>';
+  }
+
+  contents += '</configuration>\n';
+
+  return contents;
+}
+
+function formatPackageSourceElement(
+  registryInfo: ParsedRegistryUrl,
+  name: string
+): string {
+  let element = `<add key="${name}" value="${registryInfo.feedUrl}" `;
+
+  if (registryInfo.protocolVersion) {
+    element += `protocolVersion="${registryInfo.protocolVersion}" `;
+  }
+
+  return  `${element}/>\n`;
+}
+
+function formatPackageSourceCredentialElement(
+  credential: PackageSourceCredential
+): string {
+  const escapedName = credential.name.replace(
+    /(?!(\d|\w|-|\.))./g,
```

Comment:
re2 doesn't support lookahead

https://github.com/uhop/node-re2#lookahead-assertions

Reply:
Bummer, let me try something else then :)

Reply:
I worked around it by looping through the name and escaping any of the illegal characters:

https://github.com/renovatebot/renovate/pull/25052/commits/5350051f8c2caecc0e0538a1a5e55b79c7828116#diff-d84a4bda73ab057b05da33d67ff2ae3bfea6c8ed3690f02951730faa353a63f2R115-R133